pylint/checkers/unicode.py

```diff
@@ -0,0 +1,701 @@
+else:
+    from typing_extensions import Protocol
+
+if sys.version_info[:2] >= (3, 7):
+    # Once we require at least Python 3.7, we can remove this lines
+    Py37Str = str
+else:
+
+    class Py37Str(str):
+        # Allow Python 3.6 compatibility
+        def isascii(self: str) -> bool:
+            return all("\u0000" <= x <= "\u007F" for x in self)
+
+
+_StrLike = TypeVar("_StrLike", str, bytes)
+
+
+class _BadChar(NamedTuple):
+    """Representation of an ASCII char considered bad"""
+
+    name: str
+    unescaped: str
+    escaped: str
+    code: str
+    help_text: str
+
+    def description(self) -> str:
+        """Used for the detailed error message description"""
+        return (
+            f"Invalid unescaped character {self.name}, "
+            f'use "{self.escaped}" instead.'
+        )
+
+    def human_code(self) -> str:
+        """Used to generate the human readable error message"""
+        return f"invalid-character-{self.name}"
+
+
+def line_length(line: _StrLike, codec: str) -> int:
+    """Get the length of a string like line as displayed in an editor"""
+    codec = _normalize_codec_name(codec)
+
+    if isinstance(line, bytes):
+        decoded = _remove_bom(line, codec).decode(codec, "replace")
+    else:
+        decoded = line
+
+    stripped = decoded.rstrip("\n")
+
+    if stripped != decoded:
+        stripped = stripped.rstrip("\r")
+
+    return len(stripped)
+
+
+def _map_positions_to_result(
+    line: _StrLike,
+    search_dict: Dict[_StrLike, _BadChar],
+    new_line: _StrLike,
+    byte_str_length: int = 1,
+) -> Dict[int, _BadChar]:
+    """Get all occurrences of search dict keys within line
+
+    Ignores Windows end of line and can handle bytes as well as string.
+    Also takes care of encodings for which the length of an encoded code point does not
+    default to 8 Bit.
+    """
+
+    result: Dict[int, _BadChar] = {}
+    ignore_pos: Optional[int]
+    for search_for, char in search_dict.items():
+        if search_for not in line:
+            continue
+
+        # Special Handling for Windows '\r\n'
+        if char.unescaped == "\r" and line.endswith(new_line):
+            ignore_pos = len(line) - 2 * byte_str_length
+        else:
+            ignore_pos = None
+
+        start: int = 0
+        pos: int = line.find(search_for, start)
+        while pos > 0:
+            if pos != ignore_pos:
+                # Calculate the
+                col = int(pos / byte_str_length)
+                result[col] = char
+            start = pos + 1
+            pos = line.find(search_for, start)
+
+    return result
+
+
+UNICODE_BOMS: Dict[str, bytes] = {
+    "utf-8": codecs.BOM_UTF8,
+    "utf-16": codecs.BOM_UTF16,
+    "utf-32": codecs.BOM_UTF32,
+    "utf-16le": codecs.BOM_UTF16_LE,
+    "utf-16be": codecs.BOM_UTF16_BE,
+    "utf-32le": codecs.BOM_UTF32_LE,
+    "utf-32be": codecs.BOM_UTF32_BE,
+}
+
+
+@lru_cache(maxsize=1)
+def get_bom_sorted() -> Dict[bytes, str]:
```

Comment:
Can't we just initialise an `OrderedDict` and add a comment/docstring that they are sorted by size? This seems a bit overcomplicated imo.

Reply:
Something like:

```python
ENDIAN_BOMS = OrderedDict(
    [
        ("utf-8", codecs.BOM_UTF8),
        ("utf-16le", codecs.BOM_UTF16_LE),
        ("utf-16be", codecs.BOM_UTF16_BE),
        ("utf-32le", codecs.BOM_UTF32_LE),
        ("utf-32be", codecs.BOM_UTF32_BE),
    ]
)
UNICODE_BOMS = {"utf-16": codecs.BOM_UTF16, "utf-32": codecs.BOM_UTF32, **ENDIAN_BOMS}
```

Although I'm not sure if the order is as expected.